R/expr-array.R

```diff
@@ -0,0 +1,416 @@
+#'   shift_by_expr = pl$col("values")$arr$shift(pl$col("idx")),
+#'   shift_by_lit = pl$col("values")$arr$shift(2)
+#' )
+expr_arr_shift <- function(n = 1) {
+  self$`_rexpr`$arr_shift(as_polars_expr(n, as_lit = TRUE)$`_rexpr`) |>
+    wrap()
+}
+
+
+#' Convert an Array column into a List column with the same inner data type
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' df <- pl$DataFrame(
+#'   a = list(c(1, 2), c(3, 4))
+#' )$cast(pl$Array(pl$Int8, 2))
+#'
+#' df$with_columns(
+#'   list = pl$col("a")$arr$to_list()
+#' )
+expr_arr_to_list <- function() {
+  self$`_rexpr`$arr_to_list() |>
+    wrap()
+}
+
+
+# TODO-REWRITE: implement this
+# #' Convert array to struct
+# #'
+# #' @inheritParams expr_list_to_struct
+# #'
+# #' @inherit as_polars_expr return
+# #' @examples
+# #' df <- pl$DataFrame(
+# #'   values = list(1:3, c(2L, NA, 5L))
+# #' )$cast(pl$Array(pl$Int32, 3))
+# #' df$with_columns(
+# #'   struct = pl$col("values")$arr$to_struct()
+# #' )
+# #'
+# #' # pass a custom function that will name all fields by adding a prefix
+# #' df2 <- df$with_columns(
+# #'   pl$col("values")$arr$to_struct(
+# #'     fields = \(idx) paste0("col_", idx)
+# #'   )
+# #' )
+# #' df2
+# #'
+# #' df2$unnest()
+# expr_arr_to_struct <- function(fields = NULL) {
+#   self$`_rexpr`$arr_to_struct(fields) |>
+#     wrap()
+# }
+
+#' Count how often a value occurs in every sub-array
+#'
+#' @param element An Expr or something coercible to an Expr that produces a
+#' single value.
+#' @inherit as_polars_expr return
+#'
+#' @examples
+#' df <- pl$DataFrame(
+#'   values = list(c(1, 2), c(1, 1), c(2, 2))
+#' )$cast(pl$Array(pl$Int64, 2))
+#' df$with_columns(number_of_twos = pl$col("values")$arr$count_matches(2))
+expr_arr_count_matches <- function(element) {
+  self$`_rexpr`$arr_count_matches(as_polars_expr(element, as_lit = TRUE)$`_rexpr`) |>
+    wrap()
+}
+
+#' Explode array in separate rows
+#'
+#' Returns a column with a separate row for every array element.
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' df <- pl$DataFrame(
+#'   a = list(c(1, 2, 3), c(4, 5, 6))
+#' )$cast(pl$Array(pl$Int64, 3))
+#' df$select(pl$col("a")$arr$explode())
+expr_arr_explode <- function() {
+  self$`_rexpr`$explode() |>
+    wrap()
+}
+
+#' Get the first value of the sub-arrays
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' df <- pl$DataFrame(
+#'   a = list(c(1, 2, 3), c(4, 5, 6))
+#' )$cast(pl$Array(pl$Int64, 3))
+#' df$with_columns(first = pl$col("a")$arr$first())
+expr_arr_first <- function() {
+  self$`_rexpr`$arr_get(pl$lit(0)$`_rexpr`, null_on_oob = TRUE) |>
+    wrap()
+}
+
+#' Get the last value of the sub-arrays
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' df <- pl$DataFrame(
+#'   a = list(c(1, 2, 3), c(4, 5, 6))
+#' )$cast(pl$Array(pl$Int64, 3))
+#' df$with_columns(last = pl$col("a")$arr$last())
+expr_arr_last <- function() {
+  self$`_rexpr`$arr_get(pl$lit(-1)$`_rexpr`, null_on_oob = TRUE) |>
+    wrap()
+}
```

Comment:
Why not just use `self$arr$get()`?
https://github.com/pola-rs/polars/blob/be5a4b453ca74bb5e0c782c292e9fca6aa60b203/py-polars/polars/expr/array.py#L484-L507

Reply:
It fails with 
```
Error in `df$with_columns()`:
! Evaluation failed in `$with_columns()`.
Caused by error:
! Evaluation failed in `$with_columns()`.
Caused by error in `pl$col("a")$arr$first()`:
! Evaluation failed in `$first()`.
Caused by error in `self$arr`:
! $ - syntax error: `arr` is not a member of
  this polars object
```

Reply:
Ah ok but `self$get()` works